engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/ForeignEvalNode.java

```diff
@@ -0,0 +1,96 @@
+package org.enso.interpreter.epb;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import com.oracle.truffle.api.CallTarget;
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import com.oracle.truffle.api.frame.FrameDescriptor;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropException;
+import com.oracle.truffle.api.nodes.RootNode;
+import com.oracle.truffle.api.source.Source;
+
+final class ForeignEvalNode extends RootNode {
+  private final Source code;
+  private @Child ForeignFunctionCallNode foreign;
+  private @CompilationFinal ForeignParsingException parseException;
+  private final String[] argNames;
+
+  /**
+   * Creates a new instance of this node
+   *
+   * @param language the current language instance
+   * @param code the result of parsing EPB code
+   * @param arguments argument names allowed in the function body
+   * @return an instance of this node
+   */
+  static ForeignEvalNode build(
+          EpbLanguage language, Source code, List<String> arguments) {
+    return new ForeignEvalNode(language, code, arguments);
+  }
+
+  ForeignEvalNode(EpbLanguage language, Source code, List<String> arguments) {
+    super(language, new FrameDescriptor());
+    this.code = code;
+    argNames = arguments.toArray(new String[0]);
+  }
+
+  public Object execute(VirtualFrame frame) {
+    ensureParsed();
+    if (foreign != null) {
+      try {
+        var toRet = foreign.execute(frame.getArguments());
+        return toRet;
+      } catch (InteropException ex) {
+        throw new ForeignParsingException(ex.getMessage(), this);
+      }
+    } else {
+      CompilerDirectives.transferToInterpreter();
+      throw parseException;
+    }
+  }
+
+  private void ensureParsed() throws IllegalStateException {
+    if (foreign == null) {
+      CompilerDirectives.transferToInterpreterAndInvalidate();
+      String truffleLangId = EpbLanguage.truffleId(code);
+      var context = EpbContext.get(this);
+      var installedLanguages = context.getEnv().getInternalLanguages();
+      if (!installedLanguages.containsKey(truffleLangId)) {
+        this.parseException = new ForeignParsingException(truffleLangId, installedLanguages.keySet(), this);
+      } else {
+        switch (truffleLangId) {
+          case "js" -> parseJs();
```

Comment:
kind of the reason why I would prefer to have constants

Reply:
I assure you that changing the text to some other value doesn't pass CI. I tried.